pennylane/transforms/decompositions/controlled_decompositions.py

```diff
@@ -0,0 +1,108 @@
+
+
+def ctrl_decomp_zyz(target_operation: Operator, control_wires: Wires):
+    """Decompose the controlled version of a target single-qubit operation
+
+    This function decomposes a controlled single-qubit target operation using the
+    decomposition defined in section 5 of
+    `Barenco et al. (1995) <https://arxiv.org/abs/quant-ph/9503016>`_.
+
+    .. warning:: This method will add a global phase for target operations that do not
+        belong to the SU(2) group.
+
+    Args:
+        target_operation (~.operation.Operator): the target operation to decompose
+        control_wires (~.wires.Wires): the control wires of the operation
+
+    Returns:
+        list[Operation]: the decomposed operations
+
+    Raises:
+        ValueError: if `target_operation` is not a single-qubit operation
+
+    **Example**
+
+    .. code-block:: python
+
+        # We can create a controlled operation using `qml.ctrl`, or by creating the
+        # decomposed controlled version of using `qml.ctrl_decomp_zyz`.
+        op = qml.RX(0.123, wires=2)
+        ctrl_op = qml.ctrl(op, [0, 1])
+        ctrl_decomp = qml.ops.ctrl_decomp_zyz(op, [0, 1])
+
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def expected_circuit():
+            for i in range(3):
+                qml.Hadamard(i)
+            qml.apply(ctrl_op)
+            return qml.probs()
+
+        @qml.qnode(dev)
+        def decomp_circuit():
+            for i in range(3):
+                qml.Hadamard(i)
+            for op in ctrl_decomp:
+                qml.apply(op)
+            return qml.probs()
+
+    Measurements on both circuits will give us the same results:
+
+    >>> expected_circuit()
+    tensor([0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125], requires_grad=True)
+    >>> decomp_circuit()
+    tensor([0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125], requires_grad=True)
+
+    """
+    # TODO: Add support for general unitaries, not just SU(2) matrices
+    if len(target_operation.wires) != 1:
+        raise ValueError(
+            "The target operation must be a single-qubit operation, instead "
+            f"got {target_operation.__class__.__name__}."
+        )
+
+    target_wire = target_operation.wires
+
+    if len(control_wires) > 2:
+        control_gate = qml.MultiControlledX(wires=control_wires + target_wire)
+    elif len(control_wires) == 2:
+        control_gate = qml.Toffoli(wires=control_wires + target_wire)
+    else:
+        control_gate = qml.CNOT(wires=control_wires + target_wire)
+
+    try:
+        phi, theta, omega = target_operation.single_qubit_rot_angles()
+    except NotImplementedError:
+        zyz_decomp = qml.transforms.zyz_decomposition(qml.matrix(target_operation), target_wire)[0]
+        phi, theta, omega = zyz_decomp.single_qubit_rot_angles()
+
+    return [
+        qml.RZ(phi, wires=target_wire),
+        qml.RY(theta / 2, wires=target_wire),
+        control_gate,
```

Comment:
Here's a little piece of developer knowledge about an internal peculiarity of pennylane.

`Operator.expand` is constructed by "queuing" `Operator.decomposition`. Basically treating it like a quantum function.

```
with qml.queuing.AnnotatedQueue() as q:
    op.decomposition()
```

Operators are queued when they are initialized.  So if we want an operation to be after another one in the queue, we need to initialize it afterward. This also means that if we want the "same" operation multiple times in the `AnnotatedQueue`, we need to either reinitialize it or use `qml.apply` to copy and queue the original operation.

I'm not a huge fan of this "implicit assumption" we make everywhere and I am pushing to minimize the influence of queuing in PennyLane, but it's a long process.

Long story short, we need to instantiate `control_gate` here and on line 106.

Reply:
fly-by comment - this planned transition from queuing a decomposition to creating a script from one might make things messy. My gut instinct here was to just use `qml.apply(control_gate)` to fix this issue, but if someone calls this function whilst not in a queuing context, it will raise an error (which I don't think it should)

Reply:
Thanks for the insight! I think the simplest course of action is to replace `control_gate` with `qml.MultiControlledX` as that can be decomposed into `CNOT` or `Toffoli` for 1 or 2 control wires respectively as @albi3ro pointed out. That takes away both of the issues that have been brought up.